### PR TITLE
Bug fix: implement onLoad event calling when reuseMaps is set to true

### PR DESCRIPTION
**Pull request by hijiangtao**
## Description and resolution

According to the discussion in #699 , there's a TODO inside mapbox.js, which is calling onLoad event again when map instance is re-created again with its' property `reuseMaps` set to true.

However, we need to handle two situations here:

1. We need to `setStyle` with customized `mapStyle`: In this situation, we need to wait until style is loaded, and then call `onLoad` event handler; we choose `.once` API here to listen `style.load` event;
2. We don't need to `setStyle` with customized `mapStyle`: In this situation, we can directly generate an `event` object and pass it to `props.onLoad` method and call it right now;

## Tests and documentation

* Test: `npm run test` passed with 222 tests.
* Documentation: No need to update document since there's not API change with this commit;

### CODE REVIEW

File: src/mapbox/mapbox.js [hunk 1]
```diff
@@ -236,13 +236,18 @@ export default class Mapbox {
...
-      // Update style
+      // Update style and call onload again
       if (props.mapStyle) {
         this._map.setStyle(props.mapStyle);
-      }
 
-      // TODO - need to call onload again, need to track with Promise?
-      props.onLoad();
+        // call onload event handler after style fully loaded when style needs update
+        this._map.once('style.load', props.onLoad);
```
Comment (Pessimistress) [inline src/mapbox/mapbox.js]: Do we know for sure that this event will fire even if `props.mapStyle` is the same as the last style set?

**Comment by hijiangtao** [comment]
I also unified inputs of `props.onLoad` event to one object (with `type` property to be `load`), though in some situation, the event type maybe `styledata` rather than `load`.

That's because developers define `onLoadHandler` method inside `onLoad` property of react-ma-p-gl, so they would always want `event` object that passed to their `onLoadHandler` should be the same, which should always point to a `load` event.

It may not be a perfect solution but I think it's the best one right now, a better solution is to have a more precise event types (pretty the same as https://github.com/uber/react-map-gl/issues/683#issuecomment-457804805 ), then we can handler `load` event and `styledata` event separately.

A pseudocode should like this inside `_create` method (suppose there's a handler called styleLoad in props):

```
      ...

	  // Step3: update style and call onload again
      if (props.mapStyle) {
        this._map.setStyle(props.mapStyle, {
          diff: true
        });

        // call onload event handler after style fully loaded when style needs update
        (this._map.isStyleLoaded() ?
          props.styleLoad() : this._map.once('styledata', props.styleLoad))();
      }

	  props.onLoad({
        type: 'load',
        target: this._map
      });

```

File: src/mapbox/mapbox.js [hunk 2]
```diff
@@ -236,13 +236,23 @@ export default class Mapbox {
...
+      });
+
       if (props.mapStyle) {
-        this._map.setStyle(props.mapStyle);
+        this._map.setStyle(props.mapStyle, {
+          diff: true
+        });
+
+        // call onload event handler after style fully loaded when style needs update
+        (this._map.isStyleLoaded() ?
```
Comment (Pessimistress) [inline src/mapbox/mapbox.js]: This syntax is rather odd. Just use `if`?

Comment (hijiangtao) [reply src/mapbox/mapbox.js]: I got a warning of **Method '_create' has too many statements (26). Maximum allowed is 25.** if I use if statement, that’s why I choose this one syntax.

Comment (Pessimistress) [reply src/mapbox/mapbox.js]: You can add this line to the top of this method to disable the lint rule:

```js
/* eslint-disable max-statements */
```

I will refactor it after the patch.

Comment (hijiangtao) [reply src/mapbox/mapbox.js]: Yes, that's correct since it will overwrite the global settings. I can handle it by adding another commit.

**Comment by hijiangtao** [comment]
4 more implements with:

1. Enrich `examples/reuse-map` with `onLoad` event;
2. Implement TOC since reusemap example had no entrance currently;
3. Use `if` statement rather than a ternary expression;
4. Resolve conflicts from upstream;

File: examples/main/constants/toc.js [hunk 3]
```diff
@@ -76,9 +77,15 @@ export const standaloneExamples = [
...
   {
+<<<<<<< HEAD
+    path: 'reuse-map',
+    name: 'ReuseMaps',
+    component: ReuseMap
+=======
     path: 'heatmap',
     name: 'Heatmap',
     component: Heatmap
+>>>>>>> upstream/master
```
Comment (hijiangtao) [inline examples/main/constants/toc.js]: Sorry for this, I'll resolve it by adding another commit.

**Comment by Pessimistress** [comment]
Reusemap example is not on the website because it does not meet the visual quality bar. Can you remove it for now so that we can release?

**Comment by Pessimistress** [comment]
> is there any specific guidelines of the visual quality bar (I'm interested)?

Since this is a wrapper library, we try to put up examples that replicate existing [mapbox-gl-js examples](https://docs.mapbox.com/mapbox-gl-js/examples) or at least along the same lines. The reuse-map example in particular was made for performance profiling/debugging purposes so we never bothered to make it look good.

**Pull request merged**